chain/chain/src/chain.rs

```diff
@@ -2898,33 +2930,56 @@ impl<'a> ChainUpdate<'a> {
         let protocol_version =
             self.runtime_adapter.get_epoch_protocol_version(block.header().epoch_id())?;
 
+        let prev_hash = block.header().prev_hash();
+        let will_shard_layout_change = self.runtime_adapter.will_shard_layout_change(prev_hash)?;
         for (shard_id, (chunk_header, prev_chunk_header)) in
             (block.chunks().iter().zip(prev_block.chunks().iter())).enumerate()
         {
             let shard_id = shard_id as ShardId;
-            let care_about_shard = match mode {
-                ApplyChunksMode::ThisEpoch => self.runtime_adapter.cares_about_shard(
-                    me.as_ref(),
-                    &block.header().prev_hash(),
-                    shard_id,
-                    true,
-                ),
-                ApplyChunksMode::NextEpoch => Chain::should_catch_up_shard(
-                    self.runtime_adapter.clone(),
-                    me,
-                    &block.header().prev_hash(),
-                    shard_id,
-                ),
+            let cares_about_shard_this_epoch =
+                self.runtime_adapter.cares_about_shard(me.as_ref(), prev_hash, shard_id, true);
+            let cares_about_shard_next_epoch =
+                self.runtime_adapter.will_care_about_shard(me.as_ref(), prev_hash, shard_id, true);
+            let update_shard_state = match mode {
+                // next epoch's shard states are not ready, only update this epoch's shards
+                ApplyChunksMode::NotCaughtUp => cares_about_shard_this_epoch,
+                // update both this epoch and next epoch
+                ApplyChunksMode::IsCaughtUp => {
+                    cares_about_shard_this_epoch || cares_about_shard_next_epoch
+                }
+                // catching up next epoch's shard states, do not update this epoch's shard state
+                // since it has already been updated through ApplyChunksMode::NotCaughtUp
+                ApplyChunksMode::CatchingUp => {
+                    !cares_about_shard_this_epoch && cares_about_shard_next_epoch
+                }
```

Comment:
I am a bit confused by this. My understanding is that even if a node tracks the shard in the epoch before the split, it will still go through this phase due to the time it takes to split the state. However, `cares_about_shard_this_epoch` is true in this case.

Reply:
We are taking care of this outside of branch that is gated by update_shard_state

Reply:
yep, that's right